design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+###  <a name="eval">Evaluation Metrics</a>
+
+1. Developers should be able to continue using the OpenGL renderer with this refactor.
+2. The resulting binary from the modularized refactor should not increase more than 5% when compiled.
+3. The modularization refactor should be visually identical to the current implementation when using OpenGL. All existing render tests should pass, and developers that continue to use this library with an OpenGL renderer should not experience any visual degradation such as artifacts.
+4. After refactor, the startup time for applications built with this library should not increase. This will be measured by the startup time of [this](https://github.com/maplibre/maplibre-gl-native/blob/main/platform/ios/platform/ios/iosapp%20UITests/iosapp_UITests.swift) iOS test app.
+5. When implemented with a given style and set of tilesets, the compiled library should not decrease the rendering frame rate.
+6. This refactor will not introduce data loading or other bottlenecks that will negatively impact rendering performance.
+7. Developers currently using this library will not see any breaking API changes, and will be able to seamlessly upgrade versions.
+
+## Proposed Change
+
+### Shader(Program) Representation
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+We need an external representation for programs (shaders) so they can be added or replaced by developers.  
+
+Internally to the toolkit shaders are called Programs, but that's far too confusing at this level.  Are we referring to a GPU program or a program the user writes?  Best to stick with the term shader.
+
+The shaders need a representation visible outside the toolkit and we need to change the way they’re handled inside the toolkit.  We need to be able to replace them, add brand new ones, and allow developers to control their shaders.
+
+#### The way it is now
+
+Individual shaders are represented by an object class with massive amounts of template logic and a minimum of in-line comments.  They’re opaque from the outside and only controllable through vector tile data and styles. 
+
+The [render_raster_layer](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/layers/render_raster_layer.cpp), as an example, asks for the instantiation of (eventually) RasterProgram.  Rather than having the source for the program, that then pokes into a compressed chunk of memory that contains the source, which is then uncompressed and fed into OpenGL for compilation.
+
+All of this makes sense in context, but is completely inaccessible outside the toolkit and rather incomprehensible when making changes.
+
+#### Required changes
+To let developers control their own shaders we’ll need to:
+- Allow shaders to be instantiated with shader source (one option).
+- Allow shaders to be instantiated by reference to unique names in pre-compiled file (Metal does this).
+- Set uniform values by name
+- Set uniform structs by name (Metal uses structs)
```

Comment:
It can.  If you're looking for a good headache I highly recommend reading up on how you have to encode those UBOs.  :-)

But to your general point, yes, we'd want to let the various rendering implementations decide how to do this.

Reply:
> It can. If you're looking for a good headache I highly recommend reading up on how you have to encode those UBOs. :-)

When you say encode, do you mean memory layout qualifiers? std430 layout rules are pretty well documented. Even if forced to use std140, I'm a fan of inserting explicit padding in buffer object structs to satisfy the 16 byte alignment.

I'd be more curious about UBO performance on lower end hardware or with more exotic device drivers.

Reply:
Yup, the layout.  We looked into it but decided it wasn't worth the effort.  On Android we still spend too much time wiring up uniforms every frame.